internal/integrationtest/util.go

```diff
@@ -0,0 +1,255 @@
+	"go.uber.org/yarpc/peer/roundrobin"
+)
+
+// TransportSpec specifies how to create test clients and servers for a transport.
+type TransportSpec struct {
+	NewServerTransport func(t *testing.T, addr string) peer.Transport
+	NewClientTransport func(t *testing.T) peer.Transport
+	NewInbound         func(x peer.Transport, addr string) transport.Inbound
+	NewUnaryOutbound   func(x peer.Transport, pc peer.Chooser) transport.UnaryOutbound
+	Identify           func(addr string) peer.Identifier
+	Addr               func(x peer.Transport, ib transport.Inbound) string
+}
+
+// Test runs reusable tests with the transport spec.
+func (s TransportSpec) Test(t *testing.T) {
+	t.Run("reuse connection with round robin", s.TestReuseConnectionWithRoundRobin)
+	t.Run("backoff reconnection with round robin", s.TestBackoffWithRoundRobin)
+	t.Run("reconnect using round robin", s.TestReconnectWithRoundRobin)
+	t.Run("lose patience connecting with round robin", s.TestLosePatienceWithRoundRobin)
+}
+
+// NewClient returns a running dispatcher and a raw client for the echo
+// procedure.
+func (s TransportSpec) NewClient(t *testing.T, addrs []string) (*yarpc.Dispatcher, raw.Client) {
+	// Convert peer addresses into peer identifiers for a peer list.
+	ids := make([]peer.Identifier, len(addrs))
+	for i, addr := range addrs {
+		ids[i] = s.Identify(addr)
+	}
+
+	x := s.NewClientTransport(t)
+
+	pl := roundrobin.New(x)
+	pc := peerbind.Bind(pl, peerbind.BindPeers(ids))
+	ob := s.NewUnaryOutbound(x, pc)
+	d := yarpc.NewDispatcher(yarpc.Config{
+		Name: "client",
+		Outbounds: yarpc.Outbounds{
+			"service": transport.Outbounds{
+				ServiceName: "service",
+				Unary:       ob,
+			},
+		},
+	})
+	require.NoError(t, d.Start(), "start client dispatcher")
+	c := raw.New(d.ClientConfig("service"))
+	return d, c
+}
+
+// NewServer creates an echo server using the given inbound from any transport.
+func (s TransportSpec) NewServer(t *testing.T, addr string) (*yarpc.Dispatcher, string) {
+	x := s.NewServerTransport(t, addr)
+	ib := s.NewInbound(x, addr)
+
+	d := yarpc.NewDispatcher(yarpc.Config{
+		Name:     "service",
+		Inbounds: yarpc.Inbounds{ib},
+	})
+
+	handle := func(ctx context.Context, req []byte) ([]byte, error) {
+		return req, nil
+	}
+
+	d.Register(raw.Procedure("echo", handle))
+	require.NoError(t, d.Start(), "start server dispatcher")
+
+	return d, s.Addr(x, ib)
+}
+
+// TestLosePatienceWithRoundRobin is a test that any transport can apply to
+// exercise a transport dropping connections if the transport is stopped before
+// a pending request can complete.
+func (s TransportSpec) TestLosePatienceWithRoundRobin(t *testing.T) {
+	addr := "127.0.0.1:31172"
+	var wg sync.WaitGroup
+
+	client, c := s.NewClient(t, []string{addr})
+
+	wg.Add(1)
+	go func() {
+		ctx := context.Background()
+		ctx, cancel := context.WithTimeout(ctx, 50*time.Millisecond)
+		defer cancel()
+		assert.Error(t, Call(ctx, c))
+		wg.Done()
+	}()
+
+	time.Sleep(10 * time.Millisecond)
```

Comment:
I can copy pasta the mock clocks I wrote for ratelimit https://github.com/uber-go/ratelimit/tree/master/internal/clock.

However, this strategy was fickle even for ratelimit tests where the only delays were introduced by the clock, not the arbitrary delays introduced by the networking stack. I could mock all these things, but it would defeat the purpose of the integration test.

I’m investigating today how much it would cost to synchronize this test deterministically. In the worst case, I’ll make these tests pessimistic enough that they’re very unlikely to flap, and document how I arrived at these constants in the test.